### PR TITLE
Fix leaks in openssl/integer.hpp and refactoring file

### PR DESCRIPTION
**Pull request by Efrat1**
(1) fixed leak in operator=
(2) fixed leak in toHexString()
(3) convert BIGNUM num_ to unique_ptr

* **Problem Overview**  
  Two leaks and raw pointer usage.
* **Testing Done**  
  Apollo and ASAN

### CODE REVIEW

**Comment by WildFireFlum** [comment]
Close duplicate PR?

File: util/include/crypto/openssl/integer.hpp [hunk 1]
```diff
@@ -24,121 +24,121 @@
...
-    if (add_prefix) return std::string("0x") + BN_bn2hex(num_);
-    return BN_bn2hex(num_);
+    std::string result{UniqueOpenSSLString{BN_bn2hex(num_.get())}.get()};
+    if (add_prefix) {
+      result = std::string("0x") + result;
+    }
+    return result;
   }
-  std::string toDecString() const { return BN_bn2dec(num_); }
+  std::string toDecString() const { return BN_bn2dec(num_.get()); }
```
Comment (toly-kournik) [inline util/include/crypto/openssl/integer.hpp:138]: Shouldn't it use UniqueOpenSSLString as well?

Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp:138]: Thanks, I'll fix it on the other PR to master (https://github.com/vmware/concord-bft/pull/2861). Then when it's ready will create new PR to [v0.16.x](https://github.com/vmware/concord-bft/tree/v0.16.x)

**Review by glevkovich — APPROVED** [review]
LGTM

**Review by eyalrund — APPROVED** [review]

File: util/include/crypto/openssl/integer.hpp [hunk 2]
```diff
@@ -24,121 +24,134 @@
...
-    BN_CTX* ctx = BN_CTX_new();
-    if (!ctx) THROW_OPENSSL_ERROR;
-    concord::util::ScopeExit s{[&]() { BN_CTX_free(ctx); }};
+    UniqueBNCTX ctx(BN_CTX_new());
+    assertResultValid(ctx.get());
     BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_div(res, nullptr, num_, i.num_, ctx)) THROW_OPENSSL_ERROR;
-    return res;
+    assertResultValid(res);
```
Comment (WildFireFlum) [inline util/include/crypto/openssl/integer.hpp:108]: if you wrap BN_new you won't need to add this assertion manually

**Pull request merged**